### PR TITLE
Upgrade to libp2p 0.44.0

### PR DESCRIPTION
**Pull request by kpp**
polkadot companion: https://github.com/paritytech/polkadot/pull/5414
cumulus companion: https://github.com/paritytech/cumulus/pull/1221

  - What does it do?
 
Updates libp2p to 0.44.0 

Fixes https://github.com/paritytech/substrate/issues/11007
Fixes https://github.com/paritytech/substrate/issues/10370

### CODE REVIEW

File: client/network/src/peer_info.rs [hunk 1]
```diff
@@ -249,23 +246,11 @@ impl NetworkBehaviour for PeerInfoBehaviour {
 		}
 	}
 
-	fn inject_disconnected(&mut self, peer_id: &PeerId) {
-		self.ping.inject_disconnected(peer_id);
-		self.identify.inject_disconnected(peer_id);
-
-		if let Some(entry) = self.nodes_info.get_mut(peer_id) {
-			entry.info_expire = Some(Instant::now() + CACHE_EXPIRE);
```
Comment (kpp) [inline client/network/src/peer_info.rs:257]: FIXME update entry.info_expire

**Comment by tomaka** [comment]
cc @mxinden Would you mind taking a look? :pray: Since you are familiar with the breaking changes

**Comment by mxinden** [comment]
> cc @mxinden Would you mind taking a look? pray Since you are familiar with the breaking changes

On my todo list. Sorry for the delay.

**Comment by kpp** [comment]
This PR could be affected by https://github.com/paritytech/substrate/issues/10370

File: client/network/src/discovery.rs [hunk 2]
```diff
@@ -577,37 +578,27 @@ impl NetworkBehaviour for DiscoveryBehaviour {
...
 		&mut self,
 		_peer_id: &PeerId,
 		_conn: &ConnectionId,
 		_endpoint: &ConnectedPoint,
-		_handler: <Self::ProtocolsHandler as IntoProtocolsHandler>::Handler,
+		_handler: <Self::ConnectionHandler as IntoConnectionHandler>::Handler,
+		_remaining_established: usize,
 	) {
 		self.num_connections -= 1;
 		// NetworkBehaviour::inject_connection_closed on Kademlia<MemoryStore> does nothing.
```
Comment (mxinden) [inline client/network/src/discovery.rs:595]: This is no longer true.

https://github.com/libp2p/rust-libp2p/blob/b98d23263a480c77cec8f9ee9d79abc8240f2628/protocols/kad/src/behaviour.rs#L1960-L1975

**Review by mxinden — COMMENTED** [review]
Just one comment. Otherwise this looks good to me.

File: client/network/src/discovery.rs (hunk 2 above)
Comment (kpp) [reply client/network/src/discovery.rs:595]: Yes, I noticed it. We need to patch libp2p a little bit: https://github.com/libp2p/rust-libp2p/pull/2572

**Comment by kpp** [comment]
`impl NetworkBehaviour for Notifications` should be reviewed too.

**Comment by kpp** [comment]
Should `inject_listen_failure` be reviewed too?..

File: client/network/src/discovery.rs (hunk 2 above)
Comment (mxinden) [reply client/network/src/discovery.rs]: Note that https://github.com/libp2p/rust-libp2p/pull/2572 is released with libp2p v0.44.0 @kpp.

File: client/telemetry/src/transport.rs [hunk 3]
```diff
@@ -39,7 +39,14 @@ pub(crate) fn initialize_transport() -> Result<WsTrans, io::Error> {
 					future::ready(Ok::<_, io::Error>(item))
 				})
 				.try_filter(|item| future::ready(item.is_data()))
-				.map_ok(|data| data.into_bytes());
+				.map_ok(|item| {
+					let data = if let libp2p::websocket::framed::Incoming::Data(data) = item {
+						data
+					} else {
+						unreachable!("The item contains Data; qed");
```
Comment (bkchr) [inline client/telemetry/src/transport.rs:46]: How is that ensured?

File: primitives/core/Cargo.toml [hunk 4]
```diff
@@ -32,7 +32,7 @@ substrate-bip39 = { version = "0.4.4", optional = true }
 tiny-bip39 = { version = "0.8.2", optional = true }
 regex = { version = "1.5.4", optional = true }
 num-traits = { version = "0.2.8", default-features = false }
-zeroize = { version = "1.5.4", default-features = false }
+zeroize = { version = "1.4.3", default-features = false }
```
Comment (bkchr) [inline primitives/core/Cargo.toml:35]: I read your other comment, but why do we need to downgrade here as well?

**Review by bkchr — APPROVED** [review]
Most of the stuff looks reasonable, but I'm mainly trusting here @mxinden

File: client/network/Cargo.toml [hunk 5]
```diff
@@ -61,12 +61,12 @@ unsigned-varint = { version = "0.6.0", features = [
     "asynchronous_codec",
 ] }
 void = "1.0.2"
-zeroize = "1.5.4"
-libp2p = "0.40.0"
+zeroize = "1.4.3"
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/network/Cargo.toml:64]: But can we not just have different versions of zeroize?

File: primitives/core/Cargo.toml (hunk 4 above)
Comment (kpp) [reply primitives/core/Cargo.toml:35]:
```
error: failed to select a version for zeroize.
    ... required by package chacha20poly1305 v0.9.0
    ... which satisfies dependency chacha20poly1305 = "^0.9" of package snow v0.9.0
    ... which satisfies dependency snow = "^0.9.0" of package libp2p-noise v0.35.0
    ... which satisfies dependency libp2p-noise = "^0.35.0" of package libp2p v0.44.0
    ... which satisfies dependency libp2p = "^0.44.0" of package sc-authority-discovery v0.10.0-dev (/home/kpp/parity/substrate/client/authority-discovery)
    ... which satisfies path dependency sc-authority-discovery (locked to 0.10.0-dev) of package node-cli v3.0.0-dev (/home/kpp/parity/substrate/bin/node/cli)
    ... which satisfies path dependency node-cli (locked to 3.0.0-dev) of package chain-spec-builder v2.0.0 (/home/kpp/parity/substrate/bin/utils/chain-spec-builder)
versions that meet the requirements >=1, <1.5 are: 1.4.3, 1.4.2, 1.4.1, 1.4.0, 1.3.0, 1.2.0, 1.1.1, 1.1.0, 1.0.0

all possible versions conflict with previously selected packages.

  previously selected package zeroize v1.5.4
    ... which satisfies dependency zeroize = "^1.5.4" of package sc-network v0.10.0-dev (/home/kpp/parity/substrate/client/network)
    ... which satisfies path dependency sc-network (locked to 0.10.0-dev) of package beefy-gadget v4.0.0-dev (/home/kpp/parity/substrate/client/beefy)
    ... which satisfies path dependency beefy-gadget (locked to 4.0.0-dev) of package beefy-gadget-rpc v4.0.0-dev (/home/kpp/parity/substrate/client/beefy/rpc)

failed to select a version for zeroize which could resolve this conflict
```

File: client/telemetry/src/transport.rs (hunk 3 above)
Comment (kpp) [reply client/telemetry/src/transport.rs:46]: by `.try_filter(|item| future::ready(item.is_data()))`

Comment (bkchr) [reply client/telemetry/src/transport.rs:46]: Ahh I see, but this all could just be one `filter_map`. No need to have a `try_filter` and then a `map_ok`

File: client/network/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (kpp) [reply client/network/Cargo.toml:64]: I couldn't make it work.

```diff
-zeroize = "1.4.3"
+zeroize = "=1.5.4"
```

Gives the same error as below.

**Comment by tomaka** [comment]
The conflicts need to be resolved

**Comment by kpp** [comment]
@tomaka done

**Comment by kpp** [comment]
@bkchr it worked for polkadot, but not for cumulus.

**Pull request merged**